### PR TITLE
Implement authentication.

### PR DESCRIPTION
**Pull request by GregSherman**
This PR allows the user to input their key and secret in (stored with keytar) to fetch and download unlisted addons. Since we now have error messages to do with auth, I updated how they are made as well.

The key and secret input boxes are accessible from the sidebar. Ideally, they are asked on the fly when fetching an addon. However, the vscode input box api does not stay focused when clicking off the screen. This means the user cannot go and copy their code then come back without ending the process.

The `showErrorMessage` function takes in an object of the possible responses depending on the status of the response (tried to cut down on the amount of if branches there are).

The requests ask for the unlisted addon versions (filter=all_with_unlisted). Should they also ask for the deleted versions? The server code (`addons-server/src/olympia/addons/views.py:AddonVersionViewSet:check_permissions`) doesn't seem to allow both to show up in one request?

Finally, keytar gave me issues when trying to import it, and the solution was to have it as an external module in the webpack. This gave me _some_ problems with the CI tests since my systems and the CI system would not error at the same things.

### CODE REVIEW

File: src/amo/utils/addonDownload.ts [hunk 1]
```diff
@@ -3,18 +3,33 @@ import fetch from "node-fetch";
...
   if (!response.ok) {
-    await showErrorMessage(
-      `(Status ${response.status}): Could not fetch addon info.`,
-      "Request failed",
-      fetchDownloadFile,
-      [fileId]
-    );
+    const errorMessages: errorMessages = {
+      window: {
+        404: `(Status ${response.status}): Addon download file not found`,
```
Comment (eviljeff) [inline src/amo/utils/addonDownload.ts:17]:
```suggestion
        404: `(Status ${response.status}): xpi file not found`,
```

File: src/amo/utils/addonDownload.ts [hunk 2]
```diff
@@ -3,18 +3,33 @@ import fetch from "node-fetch";
...
+        401: `(Status ${response.status}): Unauthorized request for download file`,
```
Comment (eviljeff) [inline src/amo/utils/addonDownload.ts:18]:
```suggestion
        401: `(Status ${response.status}): Unauthorized request for xpi file`,
```

File: src/amo/utils/addonDownload.ts [hunk 3]
```diff
@@ -3,18 +3,33 @@ import fetch from "node-fetch";
...
+        403: `(Status ${response.status}): Forbidden request for download file`,
```
Comment (eviljeff) [inline src/amo/utils/addonDownload.ts:19]:
```suggestion
        403: `(Status ${response.status}): Inadequate permissions for xpi file`,
```

File: src/amo/utils/addonDownload.ts [hunk 4]
```diff
@@ -3,18 +3,33 @@ import fetch from "node-fetch";
...
+        other: `(Status ${response.status}): Could not fetch download file`,
```
Comment (eviljeff) [inline src/amo/utils/addonDownload.ts:20]:
```suggestion
        other: `(Status ${response.status}): Could not fetch xpi file`,
```

File: src/amo/utils/addonInfo.ts [hunk 5]
```diff
@@ -1,23 +1,37 @@
...
-      `(Status ${response.status}): Could not fetch addon info.`,
-      "Failed to fetch addon info",
-      getAddonInfo,
-      [input]
-    );
+    const errorMessages: errorMessages = {
+      window: {
+        404: `(Status ${response.status}): Addon not found`,
+        401: `(Status ${response.status}): Unauthorized request`,
+        403: `(Status ${response.status}): Forbidden request`,
```
Comment (eviljeff) [inline src/amo/utils/addonInfo.ts:21]: Similar to above, rephrase this to make it about not having enough permissions

File: src/amo/utils/addonVersions.ts [hunk 6]
```diff
@@ -2,32 +2,43 @@ import fetch from "node-fetch";
...
 
 export async function getAddonVersions(input: string, next?: string) {
   const slug: string = input.includes("/")
     ? input.split("addon/")[1].split("/")[0]
     : input;
   const url = next
     ? next
-    : `${constants.apiBaseURL}addons/addon/${slug}/versions/`;
-  const response = await fetch(url);
+    : `${constants.apiBaseURL}addons/addon/${slug}/versions?filter=all_with_unlisted`;
```
Comment (eviljeff) [inline src/amo/utils/addonVersions.ts:15]: `all_with_deleted` includes unlisted versions too.  

But we might have a problem here - the permission `ADDONS_VIEW_DELETED` isn't necessarily held by all reviewers (nor the permission for reviewing unlisted either!) and the permission check in the API will error if that filter is requested without the appropriate permission.  I'd file a follow-up for this because we'll want Andreas' opinion on what to do - if there will be scenarios when reviewers are using this plugin without `ADDONS_VIEW_DELETED` (or `ADDONS_REVIEW_UNLISTED`) then it may need to be a setting in the plugin.

File: src/amo/views/sidebarView.ts [hunk 7]
```diff
@@ -33,6 +33,22 @@ export class AssayTreeDataProvider
...
+          }
+        ),
+        new AssayTreeItem(
+          "Enter API Secret",
+          vscode.TreeItemCollapsibleState.None,
+          {
+            command: "assay.getSecret",
+            title: "Enter AMO API Secret",
+          }
+        ),
```
Comment (eviljeff) [inline src/amo/views/sidebarView.ts:51]: Having the option to set the api key/secret in the sidebar seems unnecessary when it's a one-time thing the user will need to set - can it be defined in the settings page instead?

**Review by eviljeff — CHANGES_REQUESTED** [review]

**Comment by GregSherman** [comment]
I added in a globals file so I don't have to pass around the secretStore object so much. Maybe I should look into the same for storagePath and the extension context in general?

**Review by eviljeff — APPROVED** [review]
